### PR TITLE
Make 'search' filter have a case sensitive option (for #1878)

### PR DESCRIPTION
**Pull request by navis**
First PR for druid. Welcome any comments or opinions on anything (process, code-style, etc.) 

As described in #1878, this is for supporting case-sensitive search conditions like below.
```
"filter": {
  "type": "search",
  "dimension": "product",
  "query": {
    "type": "contains",
    "value": "foo",
    "caseSensitive": true
  }
}
```

or

```
"filter": {
  "type": "search",
  "dimension": "product",
  "query": {
    "type": "fragment",
    "value": ["foo", "bar"],
    "caseSensitive": true
  }
}
```

### CODE REVIEW

File: processing/src/main/java/io/druid/query/search/search/FragmentSearchQuerySpec.java [hunk 1]
```diff
@@ -46,11 +67,21 @@ public FragmentSearchQuerySpec(
...
   public boolean accept(String dimVal)
   {
-    for (String value : values) {
-      if (dimVal == null || !dimVal.toLowerCase().contains(value.toLowerCase())) {
+    if (dimVal == null) {
+      return false;
+    }
+    final String input = caseSensitive ? dimVal : dimVal.toLowerCase();
+    for (String value : target) {
+      if (!input.contains(value)) {
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/search/search/FragmentSearchQuerySpec.java]: I think the docs are wrong here… they say fragment queries match if "any part of a dimension value contains any of the values specified" but the code matches only if all values are found.

File: processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    final String input = caseSensitive ? dimVal : dimVal.toLowerCase();
+    return input.contains(target);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] valueBytes = StringUtils.toUtf8(value);
+
+    return ByteBuffer.allocate(2 + valueBytes.length)
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: Could you put the CACHE_TYPE_ID first?

File: processing/src/main/java/io/druid/query/search/search/FragmentSearchQuerySpec.java [hunk 3]
```diff
@@ -69,8 +100,9 @@ public boolean accept(String dimVal)
       ++index;
     }
 
-    final ByteBuffer queryCacheKey = ByteBuffer.allocate(1 + valuesBytesSize)
-                                               .put(CACHE_TYPE_ID);
+    final ByteBuffer queryCacheKey = ByteBuffer.allocate(2 + valuesBytesSize)
+        .put(caseSensitive ? (byte) 1 : 0)
+        .put(CACHE_TYPE_ID);
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/search/search/FragmentSearchQuerySpec.java]: same comment about CACHE_TYPE_ID

**Comment by gianm** [comment]
test `DruidCoordinatorTest.testCoordinatorRun:349 expected:<1> but was:<0>` failed, does not seem related to this PR, bouncing.

**Comment by gianm** [comment]
This looks great @navis, thanks for the contribution!

Could you also please include docs for this change, and fill out our CLA? For the docs, I believe filters.md and searchqueryspec.md should be updated. The CLA is at: http://druid.io/community/cla.html

**Comment by vogievetsky** [comment]
Oh man, this looks awesome. I can not wait to add this to Plywood and https://github.com/implydata/druid.d.ts only thing is could you please update the docs? I need the docs because I reference them from comments in https://github.com/implydata/druid.d.ts/blob/master/druid/druid.d.ts

File: processing/src/main/java/io/druid/query/Druids.java [hunk 4]
```diff
@@ -691,6 +693,29 @@ public SearchQueryBuilder query(Map<String, Object> q)
...
 
+    public SearchQueryBuilder query(String q, boolean caseSensitive)
+    {
+      querySpec = new ContainsSearchQuerySpec(q, caseSensitive);
+      return this;
+    }
+
+    public SearchQueryBuilder query(Map<String, Object> q, boolean caseSensitive)
+    {
+      querySpec = new ContainsSearchQuerySpec((String) q.get("value"), caseSensitive);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/Druids.java]: Suggest `querySpec = new ContainsSearchQuerySpec(Preconditions.checkNotNull(q.get("value"), "no value").toString(), caseSensitive);` or similar

File: processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  private static final byte CACHE_TYPE_ID = 0x1;
+
+  private final String value;
+  private final boolean caseSensitive;
+
+  private final String target;
+
+  @JsonCreator
+  public ContainsSearchQuerySpec(
+      @JsonProperty("value") String value,
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: Is `value == null` a valid case here?

File: processing/src/test/java/io/druid/query/QueryRunnerTestHelper.java [hunk 6]
```diff
@@ -332,7 +332,7 @@ public void remove()
   {
     return new FinalizeResultsQueryRunner<T>(
         new BySegmentQueryRunner<T>(
-            segmentId, adapter.getDataInterval().getStart(),
+            adapter.getIdentifier(), adapter.getDataInterval().getStart(),
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/QueryRunnerTestHelper.java]: This should probably be in a different PR unless required for this one.

**Comment by drcrallen** [comment]
This is looking good overall. @navis can you also please make sure to apply the druid style mentioned in https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md#how-to-contribute

There are also a few files missing the copyright notice at the top.

**Comment by xvrl** [comment]
@gianm @drcrallen how about we make search case-sensitive by default?
Everything else in druid is, and we're bumping up to 0.9.x anyways. Thoughts?

File: processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  @Override
+  public boolean accept(String dimVal)
+  {
+    if (dimVal == null) {
+      return false;
+    }
+    final String input = caseSensitive ? dimVal : dimVal.toLowerCase();
+    return input.contains(target);
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: Since we are improving this code anyway, I think we should switch to using [StringUtils.containsIngoreCase](https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html#containsIgnoreCase(java.lang.String, java.lang.String)) instead of toLowerCase+contains, since lowercasing a string is not always the same for non-ascii characters.

**Comment by gianm** [comment]
@xvrl perhaps, although if so I think we should do it in a separate PR as that is a different discussion from adding a feature.

**Comment by vogievetsky** [comment]
I agree that ideally search should be case sensitive by default. Did not suggest it because did not want to break backwards compatibility.

File: processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java (hunk 2 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: sure

File: processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java (hunk 7 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: `StringUtils.containsIngoreCase` looks like checking 
```
t.toLowerCase().contains(v.toLowercase()) || t.toUpperCase().contains(v.toUpperCase())
```
I've simulated that.

File: processing/src/main/java/io/druid/query/Druids.java (hunk 4 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/Druids.java]: added null check preconditions for query/fragments

File: processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java (hunk 5 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: I don't know. Some methods are checking null and some are not. I'll update to handle nulls in all methods.

File: processing/src/test/java/io/druid/query/QueryRunnerTestHelper.java (hunk 6 above)
Comment (navis) [inline processing/src/test/java/io/druid/query/QueryRunnerTestHelper.java]: added a method not to affect existing test codes.

File: processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return value != null ? value.hashCode() : 0;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: generate hashcode again, it doesnt account for caseSensitive.

Comment (navis) [inline processing/src/main/java/io/druid/query/search/search/ContainsSearchQuerySpec.java]: done. I've changed semantic of equals() and it seemed need some debate.

**Comment by gianm** [comment]
:+1: looks good to me after [squash](https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md#github-workflow) / CLA

**Comment by navis** [comment]
Sent corporate CLA just now. Thanks.

**Comment by drcrallen** [comment]
@navis Received, thanks!

**Comment by drcrallen** [comment]
@navis if you are unable to use the intellij formatting stuff in my prior link please let me know and I'll do it on my side.

**Comment by navis** [comment]
@drcrallen I think I've applied the formatting to all of the codes affected. Any missing part? I'm good on your doing it for me. Thanks.

**Comment by fjy** [comment]
:+1:

**Comment by fjy** [comment]
@drcrallen Formatting seems fine. Merging.
